File: app/src/main/java/com/sjaindl/travelcompanion/profile/ProfileViewModel.kt
package com.sjaindl.travelcompanion.profile

import androidx.lifecycle.ViewModel
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.storage.FirebaseStorage
import com.sjaindl.travelcompanion.api.firestore.FireStoreClient
import com.sjaindl.travelcompanion.api.firestore.FireStoreConstants
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import javax.inject.Inject

@HiltViewModel
class ProfileViewModel @Inject constructor(
    private val fireStoreClient: FireStoreClient,
) : ViewModel() {
    sealed class State {
        data object Loading : State()

        data class Error(val exception: Exception?) : State()

        data object Deleted : State()

        data object Finished : State()
    }

    private val _state: MutableStateFlow<State> = MutableStateFlow(State.Loading)
    var state = _state.asStateFlow()

    private val currentUser = FirebaseAuth.getInstance().currentUser

    var userName: String = currentUser?.displayName ?: "Anonymous"

    val initials: String
        get() {
            val characters = (currentUser?.displayName ?: "TC").split(" ").mapNotNull { it.firstOrNull() }
            return if (characters.size > 1) {
                val first = characters.first()
                val last = characters.last()
                "$first$last".uppercase()
            } else {
                characters.joinToString("").uppercase()
            }
        }

    fun logout() {
        FirebaseAuth.getInstance().signOut()
    }

    fun deleteAccount() {
        val user = currentUser ?: return
        _state.value = State.Loading

        val userRef = fireStoreClient.userReference()

        deletePhotos()

        userRef.delete().addOnCompleteListener {
            user.delete().addOnCompleteListener {
                _state.value = State.Deleted
            }.addOnFailureListener {
                _state.value = State.Error(it)
            }
        }.addOnFailureListener {
            _state.value = State.Error(it)
        }
    }

    private fun deletePhotos() {
        val userRef = fireStoreClient.userReference()
        val plans = userRef.collection(FireStoreConstants.Collections.plans)

        plans.addSnapshotListener { query, error ->
            query?.documents?.forEach { planDocument ->
                val photoCollection = planDocument.reference.collection(FireStoreConstants.Collections.photos)
                photoCollection.addSnapshotListener { photoQuery, error ->
                    photoQuery?.documents?.forEach { photoDoc ->
                        (photoDoc.get("path") as? String)?.let { fileName ->
                            val storageImageRef = FirebaseStorage.getInstance().getReference(fileName)

                            storageImageRef.delete()
                                .addOnCompleteListener {
                                    if (it.isSuccessful) {
                                        _state.value = State.Finished
                                    } else {
                                        _state.value = State.Error(it.exception)
                                    }
                                }
                                .addOnFailureListener {
                                    _state.value = State.Error(it)
                                }
                        }
                    }
                }
            }
        }
    }
}

File: native/Travel Companion/Controller/Settings/ProfileViewModel.swift
//
//  ProfileViewModel.swift
//  Travel Companion
//
//  Created by Stefan Jaindl on 27.07.22.
//  Copyright © 2022 Stefan Jaindl. All rights reserved.
//

import FirebaseAuth
import FirebaseStorage
import Foundation
import UIKit

public protocol ProfileViewModelDelegate: AnyObject {
    func didUpdate(state: ProfileViewModel.State)
}

public final class ProfileViewModel {
    // MARK: Lifecycle

    public init() {
        state = State(
            user: Auth.auth().currentUser
        )
    }

    // MARK: Internal

    private(set) var state: State {
        didSet {
            delegate?.didUpdate(state: state)
        }
    }

    weak var delegate: ProfileViewModelDelegate? {
        didSet {
            delegate?.didUpdate(state: state)
        }
    }

    func logout(controller: UIViewController) {
        do {
            try Auth.auth().signOut()
        } catch {
            UiUtils.showError("signOutError".localized(), controller: controller)
        }
    }

    func item(for section: Int) -> Item? {
        guard state.items.indices.contains(section) else {
            return nil
        }
        return state.items[section]
    }
    
    func deletePhotos(controller: UIViewController) {
        let userRef = FirestoreClient.userReference()
        let plans = userRef.collection(FirestoreConstants.Collections.plans)
        
        plans.getDocuments() { query, error in
            query?.documents.forEach { planDocument in
                let photoCollection = planDocument.reference.collection(FirestoreConstants.Collections.photos)
                
                photoCollection.getDocuments() { photoQuery, photoError in
                    photoQuery?.documents.forEach { photoDoc in
                        if let fileName = photoDoc.get("path") as? String {
                            let storageImageRef = Storage.storage().reference(forURL: fileName)
                            
                            storageImageRef.delete() { error in
                                if let error = error {
                                    UiUtils.showError(error.localizedDescription, controller: controller)
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}

public extension ProfileViewModel {
    enum RowType {
        case personalInformations
        case requestAccountDeletion
    }

    enum Item: Equatable {
        case section(User?, [Item])
        case item(rowType: RowType, title: String, subTitle: String?, image: UIImage?)
    }

    struct State {
        // MARK: Lifecycle

        init(user: User?) {
            guard let user = user else {
                return
            }

            self.user = user

            items = [
                .section(user, [
                    .item(
                        rowType: .personalInformations,
                        title: "personalInformation".localized(),
                        subTitle: "nameAddressBirthday".localized(),
                        image: UIImage(named: "personal-info")
                    ),
                    .item(
                        rowType: .requestAccountDeletion,
                        title: "requestAccountDeletion".localized(),
                        subTitle: nil,
                        image: UIImage(named: "userDeleteIcon")
                    )
                ]),
            ]
        }

        // MARK: Internal

        private(set) var user: User?
        private(set) var items: [Item] = []
    }
}
